gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h

```diff
@@ -0,0 +1,110 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * NNStreamer API for Tensor_Trainer Sub-Plugins
+ * Copyright (C) 2022 Hyunil Park <hyunil46.park@samsung.com>
+ */
+/**
+ * @file  nnstreamer_plugin_api_filter.h
+ * @date  1 Dec 2022
+ * @brief Mandatory APIs for NNStreamer Trainer sub-plugins (No External Dependencies)
+ * @see https://github.com/nnstreamer/nnstreamer
+ * @author  Hyunil Park <hyunil46.park@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+#ifndef __NNS_PLUGIN_API_TRAINER_H__
+#define __NNS_PLUGIN_API_TRAINER_H__
+
+#include "tensor_typedef.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief GstTensorTrainer's properties for neural network framework (internal data structure)
+ *
+ * Internal data of GstTensorTrainer required by tensor_trainer's custom subplugin.
+ */
+typedef struct _GstTensorTrainerProperties
+{
+  GstTensorsInfo input_meta;      /**< configured input tensor info */
+  const char *model_config;     /* The configuration file path for creating model */
+  const char *model_save_path;      /* The file path to save the model */
```

Comment:
These two look too specialized for nntrainer. Make them general for all training frameworks. Be ready for model-format updates in nntrainer in the future, too.

Questions you need to ask for yourself:
1. what if a framework requires a single file for training?
2. what if a framework requires three files for training?
3. what if a framework supports multiple different formats (and different number of files accordingly)?
...

Reply:
Thank you.
About model_config, I've been thinking a lot. How about 'config_string' ?
In case of tensorflow (https://www.tensorflow.org/api_docs/python/tf/keras/models/model_from_json), use 'config' in example. param name is json_strring.
```
model = tf.keras.Sequential([
    tf.keras.layers.Dense(5, input_shape=(3,)),
    tf.keras.layers.Softmax()])
config = model.to_json()
loaded_model = tf.keras.models.model_from_json(config)
```
nntrainer use 'config'
```
  /**
   * @brief     Create and load the Network with ini configuration file.
   * @param[in] config config file path
   * @retval #ML_ERROR_NONE Successful.
   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
   */
  virtual int loadFromConfig(const std::string &config) = 0;
```  
About questions. It seems that these can be solved with GstTensorsInfo.

Currently for test.
MNIST  is also a single file for training. so after separating the file, I am going to test it by muxing it with tensor_mux and put it into tensor_trainer.

The simplest way is that separate input and label are ped into the input of the pipeline and then mux them with tensor_mux.
But....
Considering what you said.
For Various types of data(a single file for training, ....), I think it's better to process data separation outside of tensor_trainer rather than inside it.
It would be nice to if fixed GstTensorsInfo and GstTensorMemory came into tensor_trainer. 

There must be a better way, but I can't think of it yet.
We will gradually develop this.